govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/context_relevancy/context_relevancy.py

```diff
@@ -0,0 +1,202 @@
+from typing import List, Type, TypeVar
+from pydantic import BaseModel
+
+from deepeval.test_case import (
+    LLMTestCase,
+    LLMTestCaseParams,
+)
+from deepeval.metrics import BaseMetric
+from deepeval.utils import prettify_list
+from deepeval.metrics.utils import (
+    construct_verbose_logs,
+    trimAndLoadJson,
+    check_llm_test_case_params,
+    initialize_model,
+)
+from deepeval.models import DeepEvalBaseLLM
+from .template import ContextRelevancyTemplate
+from deepeval.metrics.indicator import metric_progress_indicator
+from .schema import (
+    TruthCollection,
+    InformationNeedsCollection,
+    VerdictCollection,
+    ScoreReason,
+)
+from deepeval.errors import MissingTestCaseParamsError
+
+SchemaType = TypeVar("SchemaType", bound=BaseModel)
+
+
+class ContextRelevancyMetric(BaseMetric):
+    _required_params: List[LLMTestCaseParams] = [
+        LLMTestCaseParams.INPUT,
+        LLMTestCaseParams.ACTUAL_OUTPUT,
+    ]
+
+    def __init__(
+        self,
+        threshold: float = 0.8,
+        model: str | DeepEvalBaseLLM | None = None,
+        include_reason: bool = True,
+        strict_mode: bool = False,
+        verbose_mode: bool = False,
+        evaluation_template: Type[ContextRelevancyTemplate] = ContextRelevancyTemplate,
+    ):
+        self.threshold = 1 if strict_mode else threshold
+        self.model, self.using_native_model = initialize_model(model)
+        self.evaluation_model = self.model.get_model_name()
+        self.include_reason = include_reason
+        self.strict_mode = strict_mode
+        self.verbose_mode = verbose_mode
+        self.evaluation_template = evaluation_template
+
+    async def a_measure(
+        self,
+        test_case: LLMTestCase,
+        _show_indicator: bool = True,
+        _in_component: bool = False,
+    ) -> float:
+        check_llm_test_case_params(test_case, self._required_params, self)
+        if (
+            test_case.additional_metadata is None
+            or test_case.additional_metadata.get("structured_context") is None
+        ):
+            raise MissingTestCaseParamsError(
+                "additional_metadata['structured_context']"
+                " cannot be None for ContextRelevancyMetric."
+            )
+
+        with metric_progress_indicator(
+            self,
+            async_mode=True,
+            _show_indicator=_show_indicator,
+            _in_component=_in_component,
+        ):
+            if isinstance(test_case.additional_metadata, dict):
+                structured_context = test_case.additional_metadata.get(
+                    "structured_context", []
+                )
+                retrieval_context = [
+                    ctx.to_flattened_context_content() for ctx in structured_context
+                ]
+            else:
+                retrieval_context = []
+
+            self.evaluation_cost = 0.0
```

Comment:
This seems a somewhat surprising spot to set this variable

Reply:
Yea i initially had it as part of the init, but moved it on previous feedback. Maybe it should live in the incrment cost record i.e.

```
if self.cost.defined?
  self.evaluation_cost += cost
else 
  self.evaluation_cost = cost
end
```
excuse the ruby sudocode

Wdyt?